src/opencl_bitlocker_fmt_plug.c

```diff
@@ -0,0 +1,773 @@
+	s = MIN(s, autotune_get_task_max_work_group_size(FALSE, 0, prepare_kernel));
+	s = MIN(s, autotune_get_task_max_work_group_size(FALSE, 0, final_kernel));
+	s = MIN(s, autotune_get_task_max_work_group_size(FALSE, 0, block_kernel));
+
+	return s;
+}
+
+static void release_clobj(void)
+{
+	if (hostFound[0] >= 0) {
+		HANDLE_CLERROR(clReleaseMemObject(deviceEncryptedVMK), "Release encrypted VMK");
+		HANDLE_CLERROR(clReleaseMemObject(devicePassword), "Release encrypted VMK");
+		HANDLE_CLERROR(clReleaseMemObject(devicePasswordSize), "Release encrypted VMK");
+		HANDLE_CLERROR(clReleaseMemObject(deviceFound), "Release encrypted VMK");
+		HANDLE_CLERROR(clReleaseMemObject(w_blocks_d), "Release encrypted VMK");
+		HANDLE_CLERROR(clReleaseMemObject(salt_d), "Release encrypted VMK");
+		HANDLE_CLERROR(clReleaseMemObject(padding_d), "Release encrypted VMK");
+
+		MEM_FREE(hostFound);
+		MEM_FREE(w_blocks_h);
+		MEM_FREE(numPasswordsKernel);
+	}
+}
+
+static void init(struct fmt_main *_self)
+{
+	self = _self;
+	opencl_prepare_dev(gpu_id);
+}
+
+static void reset(struct db_main *db)
+{
+	if (!autotuned) {
+
+		char build_opts[64];
+
+		snprintf(build_opts, sizeof(build_opts), "-DHASH_LOOPS=%u", HASH_LOOPS);
+
+		opencl_init("$JOHN/kernels/bitlocker_kernel.cl", gpu_id, NULL);
+
+		block_kernel =
+		    clCreateKernel(program[gpu_id], WBLOCKS_KERNEL_NAME, &cl_error);
+		HANDLE_CLERROR(cl_error,
+		               "Error creating block kernel");
+
+		prepare_kernel =
+			clCreateKernel(program[gpu_id], PREPARE_KERNEL_NAME, &cl_error);
+		HANDLE_CLERROR(cl_error, "Error creating crypt kernel");
+
+		crypt_kernel =
+			clCreateKernel(program[gpu_id], ATTACK_KERNEL_NAME, &cl_error);
+		HANDLE_CLERROR(cl_error, "Error creating crypt kernel");
+
+		final_kernel =
+			clCreateKernel(program[gpu_id], FINAL_KERNEL_NAME, &cl_error);
+		HANDLE_CLERROR(cl_error, "Error creating crypt kernel");
+
+		// Initialize openCL tuning (library) for this format.
+		opencl_init_auto_setup(SEED, HASH_LOOPS, split_events, warn,
+		                       14, self, create_clobj, release_clobj,
+		                       BITLOCKER_INT_HASH_SIZE * sizeof(unsigned int),
+		                       0, db);
+
+		// Auto tune execution from shared/included code.
+		autotune_run(self, ITERATIONS, 0,
+		             (cpu(device_info[gpu_id]) ?
+		              1000000000 : 100));
+	}
+}
+
+static void done(void)
+{
+	if (autotuned) {
+		release_clobj();
+		HANDLE_CLERROR(clReleaseKernel(block_kernel), "Release kernel W");
+		HANDLE_CLERROR(clReleaseKernel(prepare_kernel), "Release kernel A");
+		HANDLE_CLERROR(clReleaseKernel(crypt_kernel), "Release kernel A");
+		HANDLE_CLERROR(clReleaseKernel(final_kernel), "Release kernel A");
+
+		//HANDLE_CLERROR(clReleaseKernel(split_kernel), "Release kernel 2");
+		HANDLE_CLERROR(clReleaseProgram(program[gpu_id]),
+		               "Release Program");
+
+		autotuned--;
+	}
+
+}
+
+static void *get_salt(char *ciphertext)
```

Comment:
Yes, I forgot to advise you about this issue: is get_salt returns the ciphertext address, there will be a crash in formats.c. Is it a bug? Anyway, I restored the code line in get_salt to fix this issue.